Tests/Client/Unit Tests/WebSocket/WebSocketTests.swift

```diff
@@ -11,124 +10,140 @@ import XCTest
-            guard case .connectionChanged(let state) = event else {
-                return
-            }
-            
-            if case .connected = state {
-                connectedExpectation.fulfill()
-                shouldntConnectExpectation.fulfill()
-            } else if case .reconnecting = state {
-                reconnectingExpectation.fulfill()
-            }
-        }
+    func test_reconnectionFlow_withStopError_shouldntReconnect() {
+        // Setup (connect)
+        test_connectionFlow()
+                        
+        let stopError = WebSocketProviderError(reason: "test",
+                                               code: WebSocketProviderError.stopErrorCode,
+                                               providerType: WebSocketProviderMock.self,
+                                               providerError: nil)
         
-        mockProvider.failNextConnectCount = 3
-        webSocket.connect()
-        wait(for: [shouldntConnectExpectation], timeout: 2)
-        wait(for: [connectedExpectation, reconnectingExpectation], timeout: 10)
+        let originalConnectCalledCount = socketProvider.connectCalledCount
+
+        // Action
+        socketProvider.simulateDisconnect(stopError)
+        time.run(numberOfSeconds: maxReconnectTimeout + 1)
+        
+        // Assert
+        AssertAsync {
+            Assert.staysTrue(self.socketProvider.connectCalledCount == originalConnectCalledCount)
+        }
     }
-    
-    func test_webSocket_ping() {
-        // Keep Connection for 4 sec.
-        let shouldntDisconnectWithStopErrorExpectation = expectation(description: "WebSocket shouldn't recieve stop error")
-        shouldntDisconnectWithStopErrorExpectation.isInverted = true
+
+    func test_reconnectionFlow_whenDisconnectedManually_shouldntReconnect() {
+        // Setup (connect)
+        test_connectionFlow()
+        let originalConnectCalledCount = socketProvider.connectCalledCount
+
+        // Action
+        webSocket.disconnect(reason: "Tests")
+        time.run(numberOfSeconds: maxReconnectTimeout + 1)
         
-        let webSocket = WebSocket(mockProvider, options: .stayConnectedInBackground) { event in
-            if case .connectionChanged(let state) = event,
-                case .disconnected(let clientError) = state,
-                case .websocketDisconnectError(let websocketDisconnectError) = clientError,
-                let webSocketProviderError = websocketDisconnectError as? WebSocketProviderError,
-                webSocketProviderError.code == WebSocketProviderError.stopErrorCode {
-                shouldntDisconnectWithStopErrorExpectation.fulfill()
-            }
+        // Assert
+        AssertAsync {
+            Assert.staysTrue(self.socketProvider.connectCalledCount == originalConnectCalledCount)
         }
+    }
+
+    func test_pingIsSentPeriodically() {
+        // Setup (connect)
+        test_connectionFlow()
+        assert(WebSocket.pingTimeInterval == 25)
```

Comment:
I don't think this makes sense here. We shouldn't assume ping interval will be 25, and write the test with accessing this value (calculated)